### PR TITLE
[TextFields] Padded label for placeholder for outlined style.

### LINKED ISSUE #1921: [Textfields] Add horizontal padding to floating titles
floating titles should have 4dp of padding on the left and right of the text.

### PR DESCRIPTION
**Pull request by willlarche**
Closes #1921

### CODE REVIEW

File: components/TextFields/src/private/MDCPaddedLabel.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#import <UIKit/UIKit.h>
+/**
+ Deltas for padding.
+
+ horizontal is the dx and vertical is the dy.
+ */
+struct MDCSymmetricPadding {
+  CGFloat horizontal;
+  CGFloat vertical;
+};
```
Comment (ianegordon) [inline components/TextFields/src/private/MDCPaddedLabel.h]: Can we not use UIEdgeInsets?
Perhaps with a convenience method: UIEdgeInsetsWithSymetricalPadding(horizontal, vertical)

File: components/TextFields/src/private/MDCPaddedLabel.m [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+#pragma mark - UILabel Overrides
+
+- (void)drawTextInRect:(CGRect)rect {
+  [super drawTextInRect:CGRectInset(rect, self.padding.horizontal, self.padding.vertical)];
+}
+
+#pragma mark - UIView Overrides
+
+- (CGSize)intrinsicContentSize {
```
Comment (ianegordon) [inline components/TextFields/src/private/MDCPaddedLabel.m:37]: Does intrinsicContentSize work with textRectForBounds:limitedToNumberOflines: ?

File: components/TextFields/src/private/MDCTextInputCommonFundament.m [hunk 3]
```diff
@@ -19,6 +19,7 @@
 #import "MDCTextInputArt.h"
 #import "MDCTextInputBorderView.h"
 #import "MDCTextInputCommonFundament.h"
+#import "MDCPaddedLabel.h"
```
Comment (ajsecord) [inline components/TextFields/src/private/MDCTextInputCommonFundament.m]: Sort

**Review by ajsecord — CHANGES_REQUESTED** [review]
Is this a visual change, if so, can you please add a before/after screenshot?

File: components/TextFields/src/private/MDCPaddedLabel.h (hunk 1 above)
Comment (willlarche) [reply components/TextFields/src/private/MDCPaddedLabel.h]: The labels don't have 'real' insets: the left can never be different than the right. And that's by design: it's the only case needed by the class using it and it's a private class.

So I had a conversation with 3 developers asking what would be the most natural way to express these values: CGPoint, CGSize, UIEdgeInsets, Custom struct. After much deliberation, 2:3 vote for custom struct.

**Comment by willlarche** [comment]
The difference visually:

padded label

File: components/TextFields/src/private/MDCTextInputCommonFundament.m (hunk 3 above)
Comment (willlarche) [reply components/TextFields/src/private/MDCTextInputCommonFundament.m]: Done!

File: components/TextFields/src/private/MDCPaddedLabel.m (hunk 2 above)
Comment (willlarche) [reply components/TextFields/src/private/MDCPaddedLabel.m:37]: I have every reason to believe it does. That method is called to get the size the _text_ should be which is unchanged by my changes. However, if we want this to have even better support, I should add an overridden sizeThatFits:

File: components/TextFields/src/private/MDCPaddedLabel.h (hunk 1 above)
Comment (ianegordon) [reply components/TextFields/src/private/MDCPaddedLabel.h]: ☹️  Creating a custom struct that is almost, but not quite entirely, identical to an existing, commonly used struct strikes me as unfriendly.
Consider that using a UIEdgeInsets might also simplify your documentation:
/**
 The padding to be applied.

  paddingInsets are used to inset the target rect in drawTextInRect and to increase the intrinsicContentSize.
*/
I'll leave it to you.

File: components/TextFields/src/MDCTextInputControllerOutlinedField.m [hunk 4]
```diff
@@ -116,30 +128,49 @@ - (void)updatePlaceholder {
...
+                                                            constant:placeholderConstant];
     self.placeholderCenterY.priority = UILayoutPriorityDefaultHigh;
     self.placeholderCenterY.active = YES;
 
     [self.textInput.placeholderLabel setContentHuggingPriority:UILayoutPriorityDefaultHigh + 1
                                                        forAxis:UILayoutConstraintAxisVertical];
   }
   self.placeholderCenterY.constant = placeholderConstant;
+
+  struct MDCSymmetricPadding padding = ((MDCPaddedLabel *)self.textInput.placeholderLabel).padding;
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextInputControllerOutlinedField.m]: casting is where things often go wrong.
Please move all casts into there own statement / localvar to ease in parsing.  Here an below.
MDCPaddedLabel *paddedLabel = (MDCPaddedLabel *)self.textInput.placeholderLabel;

**Review by ianegordon — CHANGES_REQUESTED** [review]
Please pull the casts into their own statements.
I would argue longer about the struct if this was a public interface.

File: components/TextFields/src/MDCTextInputControllerOutlinedField.m (hunk 4 above)
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerOutlinedField.m]: This is done.

File: components/TextFields/src/private/MDCPaddedLabel.m (hunk 2 above)
Comment (willlarche) [reply components/TextFields/src/private/MDCPaddedLabel.m:37]: Added comment.

**Review by ianegordon — APPROVED** [review]

**Comment by ianegordon** [comment]
Please merge when Travis is happy.

**Review by ajsecord — APPROVED** [review]

**Pull request merged**